metrics/graphite/graphite.go

```diff
@@ -0,0 +1,335 @@
+// Package graphite implements a Graphite backend for package metrics. Metrics
+// will be emitted to a Graphite server in the plaintext protocol
+// (http://graphite.readthedocs.io/en/latest/feeding-carbon.html#the-plaintext-protocol)
+// which looks like:
+//   "<metric path> <metric value> <metric timestamp>"
+//
+// The current implementation ignores fields.
+package graphite
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math"
+	"net"
+	"sort"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/codahale/hdrhistogram"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/metrics"
+)
+
+// Emitter will keep track of all metrics and, once started,
+// will emit the metrics via the Flush method to the given address.
+type Emitter struct {
+	prefix string
+
+	addr  string
+	tcp   bool
+	conn  net.Conn
+	start sync.Once
+	stop  chan bool
+
+	mtx        sync.Mutex
+	counters   []*counter
+	histograms []*windowedHistogram
+	gauges     []*gauge
+
+	logger log.Logger
+}
+
+// NewEmitter will return an Emitter that will prefix all
+// metrics names with the given prefix. Once started, it will attempt to create
+// a TCP or a UDP connection with the given address and periodically post
+// metrics to the connection in the Graphite plaintext protocol.
+// If the provided `tcp` parameter is false, a UDP connection will be used.
+func NewEmitter(addr string, tcp bool, metricsPrefix string, logger log.Logger) *Emitter {
```

Comment:
I'd like to see this use the parameters from [net.Dial](https://golang.org/pkg/net/#Dial), namely `network, address string`, and drop `tcp bool`. Then you can have another constructor that parameterizes the dialer, which should really help with testing. 

```go
type Emitter struct {
    network string
    address string
    dial    Dialer
    // ....
}

type Dialer func(network, address string) (net.Conn, error)

func NewEmitter(network, address string, ...) *Emitter {
    return NewEmitterDial(network, address, net.Dial, ...)
}

func NewEmitterDial(network, address string, dial Dialer, ...) *Emitter {
    return &Emitter{
        network: network,
        address: address,
        dial:    dial,
        // ...
    }
}

func (e *Emitter) dial() {
    conn, err := e.dial(e.network, e.address)
    // ...
}
```

What do you think?